_studio/mfx_lib/lpla/mfx_lp_lookahead.cpp

```diff
@@ -0,0 +1,283 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "mfx_lp_lookahead.h"
+#include "mfx_h265_encode_hw.h"
+#include "mfx_h265_encode_hw_utils.h"
+#include "mfx_vpp_main.h"
+#include "libmfx_core_interface.h"
+
+#if defined (MFX_ENABLE_LP_LOOKAHEAD)
+
+#define MSDK_ALIGN16(value)             (((value + 15) >> 4) << 4)
+#define MSDK_MEMCPY_VAR(dstVarName, src, count) memcpy(&(dstVarName), (src), (count))
+
+mfxStatus MfxLpLookAhead::Init(mfxVideoParam* param)
+{
+    mfxStatus mfxRes = MFX_ERR_NONE;
+    MFX_CHECK_NULL_PTR1(param);
+
+    if (m_bInitialized)
+        return MFX_ERR_NONE;
+
+    if (CheckDownScaling(*param))
+    {
+        mfxRes = CreateVpp(*param);
+        MFX_CHECK_STS(mfxRes);
+    }
+
+    m_pEnc = new MFXVideoENCODEH265_HW(m_core, &mfxRes);
+    MFX_CHECK_NULL_PTR1(m_pEnc);
+
+    // following configuration comes from HW recommendation
+    mfxVideoParam par         = *param;
+    par.AsyncDepth            = 1;
+    par.mfx.CodecId           = MFX_CODEC_HEVC;
+    par.mfx.LowPower          = MFX_CODINGOPTION_ON;
+    par.mfx.NumRefFrame       = 1;
+    par.mfx.TargetUsage       = 7;
+    par.mfx.RateControlMethod = MFX_RATECONTROL_CQP;
+    par.mfx.CodecProfile      = MFX_PROFILE_HEVC_MAIN;
+    par.mfx.CodecLevel        = MFX_LEVEL_HEVC_52;
+    par.mfx.QPI               = 30;
+    par.mfx.QPP               = 32;
+    par.mfx.QPB               = 32;
+    par.mfx.NumSlice          = 1;
+
+    if (m_bNeedDownscale)
+    {
+        par.mfx.FrameInfo.CropX = 0;
+        par.mfx.FrameInfo.CropY = 0;
+        par.mfx.FrameInfo.CropW = m_dstWidth;
+        par.mfx.FrameInfo.CropH = m_dstHeight;
+
+        par.mfx.FrameInfo.Width = MSDK_ALIGN16(m_dstWidth);
+        par.mfx.FrameInfo.Height= MSDK_ALIGN16(m_dstHeight);
+    }
+
+    //create the bitstream buffer
+    memset(&m_bitstream, 0, sizeof(mfxBitstream));
+    mfxU32 bufferSize = param->mfx.FrameInfo.Width * param->mfx.FrameInfo.Height * 3/2;
+    m_bitstream.Data = new mfxU8[bufferSize];
```

Comment:
I suggest to use simple wrapper for bitstream object (or keep memory for bitstream in some ```std::unique_ptr```)